### PR TITLE
[libc++][doc] Updates module build instructions.

### PR DESCRIPTION
**Pull request by mordante**
CMake has landed experimental support for using the Standard modules. This will be part of the CMake 3.30 release. This updates the build instructions to use modules with CMake.

The changes have been tested locally.

### CODE REVIEW

File: libcxx/docs/Modules.rst [hunk 1]
```diff
@@ -214,6 +219,62 @@ Building this project is done with the following steps, assuming the files
...
+  )
+
+  #
+  # Set language version used
+  #
+
+  set(CMAKE_CXX_STANDARD 23)
+  set(CMAKE_CXX_STANDARD_REQUIRED YES)
+  # This seems to be required to be on.
+  set(CMAKE_CXX_EXTENSIONS ON)
```
Comment (mordante) [inline libcxx/docs/Modules.rst:257]: @mathstuf Is requiring the `CMAKE_CXX_EXTENSIONS` a known limitation?

Comment (mathstuf) [reply libcxx/docs/Modules.rst:257]: CMake doesn't care about this in the relevant codepaths, so if it is, it's a compiler thing.

File: libcxx/docs/Modules.rst [hunk 2]
```diff
@@ -105,9 +103,17 @@ Users need to be able to build their own BMI files.
...
-Currently this requires a local build of libc++ with modules enabled. Since
-modules are not part of the installation yet, they are used from the build
-directory. First libc++ needs to be build with module support enabled.
+Currently there are two ways to build modules
+
+  * Use a local build of modules from the build directory. This requires
+    Clang 17 or later and CMake 3.26 or later.
+
+  * Use the installed modules. This requires Clang 18.1.2 or later and
+    a recent build of CMake. The CMake changes will be part CMake 3.30. This
```
Comment (hawkinsw) [inline libcxx/docs/Modules.rst:112]:
```suggestion
    a recent build of CMake. The CMake changes will be part of CMake 3.30. This
```

File: libcxx/docs/Modules.rst [hunk 3]
```diff
@@ -214,6 +219,62 @@ Building this project is done with the following steps, assuming the files
 
              ``error: module file _deps/std-build/CMakeFiles/std.dir/std.pcm cannot be loaded due to a configuration mismatch with the current compilation [-Wmodule-file-config-mismatch]``
 
+
+Using the installed modules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+CMake has added experimental support for importing the standard modules. This
+is available in the current nightly builds and will be part of the 3.30
+release. Currently CMake only supports importing the Standard modules in C++23
```
Comment (hawkinsw) [inline libcxx/docs/Modules.rst:228]: Do we want to write "Standard" or "standard"?

**Review by hawkinsw — COMMENTED** [review]
Thank you for adding this helpful documentation! I hope that my little bit of feedback is helpful.

File: libcxx/docs/Modules.rst (hunk 1 above)
Comment (mordante) [reply libcxx/docs/Modules.rst:257]: I think it's a CMake issue, I've filed https://gitlab.kitware.com/cmake/cmake/-/issues/25916 with additional information.

File: libcxx/docs/Modules.rst (hunk 3 above)
Comment (mordante) [reply libcxx/docs/Modules.rst:228]: The official wording uses `International Standard` so I tend to capitalize; that is if I don't forget to do it ;-)

Comment (hawkinsw) [reply libcxx/docs/Modules.rst:228]: I thought so -- I am terrible with that sort of thing :-)

File: libcxx/docs/Modules.rst [hunk 4]
```diff
@@ -214,6 +219,64 @@ Building this project is done with the following steps, assuming the files
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+CMake has added experimental support for importing the Standard modules. This
+is available in the current nightly builds and will be part of the 3.30
+release. Currently CMake only supports importing the Standard modules in C++23
+and later. Enabling this for C++20 is on the TODO list of the CMake
+developers.
+
+The example uses the same ``main.cpp`` as above. It uses the following
+``CMakeLists.txt``
```
Comment (ldionne) [inline libcxx/docs/Modules.rst:233]:
```suggestion
``CMakeLists.txt``:
```

File: libcxx/docs/Modules.rst [hunk 5]
```diff
@@ -214,6 +219,64 @@ Building this project is done with the following steps, assuming the files
...
+  # This requires a recent nightly build.
+  # This will be part of CMake 3.30.0.
+  cmake_minimum_required(VERSION 3.29.0 FATAL_ERROR)
+
+  # Enables the Standard module support. This needs to be done
+  # before selecting the languages.
+  set(CMAKE_EXPERIMENTAL_CXX_IMPORT_STD "0e5b6991-d74f-4b3d-a41c-cf096e0b2508")
+  set(CMAKE_CXX_MODULE_STD ON)
+
+  project("module"
```
Comment (ldionne) [inline libcxx/docs/Modules.rst:246]:
```suggestion
  project("example"
```

**Review by ldionne — APPROVED** [review]

**Comment by ldionne** [comment]
At this point, I think we should consider installing the modules source files by default. It will still be up to the vendors whether they want to ship them. Once vendors start installing them (even if they're still experimental), this would get the ball rolling for build systems to add support for modules. It's a bit of a chicken-and-egg problem, someone needs to decide to ship something that isn't 100% ready yet. I think is has to be us.

**Pull request merged**